addons/l10n_it_stock_ddt/tests/test_ddt.py

```diff
@@ -88,3 +88,48 @@ def test_ddt_flow(self):
         self.inv2 = self.so._create_invoices()
         self.inv2.action_post()
         self.assertEqual(self.inv2.l10n_it_ddt_ids.ids, (pickx1 | pickx2).ids, 'DDTs should be linked to the invoice')
+
+    def test_ddt_flow_2(self):
+        """
+            Test that the link between the invoice lines and the deliveries linked to the invoice
+            through the link with the sale order is calculated correctly.
+        """
+        so = self.env['sale.order'].create({
+            'partner_id': self.partner_a.id,
+            'order_line': [(0, 0, {
+                                   'product_id': self.product_a.id,
+                                   'product_uom_qty': 3,
+                                   'product_uom': self.product_a.uom_id.id,
+                                   'price_unit': self.product_a.list_price,
+                                   'tax_id': self.company_data['default_tax_sale']
+                                   }
+                            )],
+            'pricelist_id': self.company_data['default_pricelist'].id,
+            'picking_policy': 'direct',
+        })
+        so.action_confirm()
+
+        # deliver partially
+        picking_1 = so.picking_ids
+        picking_1.move_lines.write({'quantity_done': 1})
+        wiz_act = picking_1.button_validate()
+        wiz = Form(self.env[wiz_act['res_model']].with_context(wiz_act['context'])).save()
+        wiz.process()
+
+        invoice_1 = so._create_invoices()
+        invoice_form = Form(invoice_1)
+        with invoice_form.invoice_line_ids.edit(0) as line:
+            line.quantity = 1.0
+        invoice_1 = invoice_form.save()
+        invoice_1.action_post()
+
+        picking_2 = so.picking_ids.filtered(lambda p: p.state != 'done')
+        picking_2.move_lines.write({'quantity_done': 2})
+        picking_2.button_validate()
+
+        invoice_2 = so._create_invoices()
+        invoice_2.action_post()
+
+        self.env.cache.invalidate()
```

Comment:
@DjamelTouati Not usre why we need a cache invalidation?

Reply:
Hi @jco-odoo
we need to invalidate the cache to ensure that the `account.move.lines` are in the order from database (_order on models) and not in the order of values currently in cache https://github.com/odoo/odoo/blob/bde2be069910596df9433df684309fe9f21a494a/addons/l10n_it_stock_ddt/models/account_invoice.py#L33
we do something similar here too: https://github.com/odoo/enterprise/blob/b8fa3bbffad83fe2749049c46e82674ec31ea007/account_reports/tests/test_tax_report.py#L383